### DISCUSSION TITLE
Unable to decompress PSD image with layer mask

### DISCUSSION THREAD

**Discussion by Ser-Gen** [ImageMagick/ImageMagick#1107] [General]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am using the latest version of ImageMagick
- [x] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Got error on psd parsing (`convert`, `identify`).

### Steps to Reproduce
1. Get file [bg_mask.zip](https://github.com/ImageMagick/ImageMagick/files/1938910/bg_mask.zip)
2. Execute `magick convert bg_mask.psd bg_mask.png`.
3. Get errors
```
convert: unable to decompress image `bg_mask.psd' @ error/psd.c/ReadPSDChannel/1425.
convert: no images defined `bg_mask.png' @ error/convert.c/ConvertImageCommand/3275.
```
4. Execute `magick convert bg_mask.psd[0] bg_mask.png`
5. No errors, get png.

### System Configuration

- ImageMagick version:
Version: ImageMagick 7.0.7-28 Q16 x64 2018-03-25
- Environment (Operating system, version and so on):
Windows 10 x64

**Comment by dlemstra** [comment]
Your PSD is corrupt and that is why it fails when you try to read all files. With `bg_mask.psd[0]` you will only read the composite image that is not corrupt and that is why you can get an output image when you do that.

**Comment by Ser-Gen** [comment]
But I can get this psd from PS CS6 and CC just like that.

![ps_solid_color](https://user-images.githubusercontent.com/1856145/39169334-1d24362a-47a0-11e8-90d1-44b46f8e8cb1.png)

![image](https://user-images.githubusercontent.com/1856145/39169349-25a587cc-47a0-11e8-96fe-93ec2f125adf.png)

Here is file that I get by this operation, same error from IM.
[bg_mask_test.zip](https://github.com/ImageMagick/ImageMagick/files/1941112/bg_mask_test.zip)

This file can be successfully parsed by Photopea.

Are you sure that my file is corrupted?

**Comment by fmw42** [comment]
This will not help, but for IM 7, the proper syntax is magick rather than magick convert or even convert.

**Comment by dlemstra** [comment]
It might be possible that we read the file incorrectly. Will need to investigate this.

**Comment by Mikluha** [comment]
I have a similar problem! there are some changes?
version: ImageMagick 6.9.9-45 Q16 x86_64

**Comment by blitline-dev** [comment]
This is causing an issue for us as well. We use multiple IM versions including 6.0, 6.9+ and 7+. None of the IM versions seem to work with these layer mask files.

**Comment by fmw42** [comment]
The psd file has a layer mask and not an alpha channel. Dirk will need to comment whether ImageMagick can handle layer masks vs alpha channels. I am unaware if he has upgraded IM for layer masks. I believe it would work if you changed the layer mask to an alpha channel.

**Comment by charlyv** [comment]
We are facing the same issue with some Photoshop files that cannot be converted to JPEG or PNG because of this error:

unable to decompress image `/mnt/data1/ibData/resources/05d4fd0517e6f7e1ee5ef12e9086f9e5/a63ad217fc5bbc5e6471374778f961ef.psd'

Example file: 

https://cdn.intelligencebank.com/us/share/vkPX/21wX/JGRJ/original/INT-Burger-Classic-Morita-Mex-101118

Any insights on the timing for a fix, or a work around solution would be greatly appreciated.

Thanks

**Comment by fmw42** [comment]
I can convert it fine using IM 6.9.10.46 or IM 7.0.8.46 Mac OSX Sierra with

`convert INT-Burger-Classic-Morita-Mex-101118.psd[0] INT-Burger-Classic-Morita-Mex-101118.png`

**Comment by charlyv** [comment]
Thanks, yes, that works for us as well, we were just hoping there would be a way to do this without using the composite image as this requires to implement more complex detection and handling logic on user upload in our interface.

Right now, this only works if I go to the file in Photoshop, and select Layer > Layer Style > Hide All Effects and save the file that way before conversion.

**Comment by fmw42** [comment]
I am not an expert on the Imagemagick code, but to my understanding Imagemagick has limited abilities with regard to advance photoshop features such as groups, layer name, extra channels, adjustment layers. Imagemagick typically can only handle simple layers. So it is not surprising to me that you would have issues.  But I will defer to the Imagemagick developers. Perhaps you should provide and example of the Imagemagick code you are trying to use and what you are trying to do with it functionally. That way it can be tested.

**Comment by marcschneider123** [comment]
I am having the same issue with attached psd including layer masks
[IMG_5246b.psd.zip](https://github.com/ImageMagick/ImageMagick/files/3359631/IMG_5246b.psd.zip)

PHP 7.3.6
imagick 3.4.4
ImageMagick 7.0.8-49 Q16 x86_64

running this:
```
<?php
$file = './IMG_5246b.psd';
$result = './IMG_5246b.jpg';

try{
	$im = new imagick($file);
	$im->writeImage($result);
	echo 'done';
} catch(\Exception $e){
	echo "\nCannot generate thumbnail : ".$e->getMessage();
}
?>
```

> unable to decompress image `./testfiles/IMG_5246b.psd' @ error/psd.c/ReadPSDChannel/1442

**Comment by fmw42** [comment]
Works fine for me by specifying the first layer, which is the flattened image using IM 7.0.10.10 Q16 Mac OSX as

`magick IMG_5246b.psd[0] x.jpg`

Try adding [0] to the end of your input file in your command above as

`$file = './IMG_5246b.psd[0]';`
